File: day-05/part-2/badouralix.js

/**
* @param {{string}} s puzzle input in string format
* @returns solution flag
*/
function run(s) {
	const offsets = s.split('\n').map(x => parseInt(x));
	let index = 0, steps = 0,
		currentOffset;

	while (index >= 0 && index < offsets.length) {
		currentOffset = offsets[index];
		offsets[index] += offsets[index] < 3 ? 1 : -1;
		index += currentOffset;
		steps++;
	}

	return steps;
}

File: day-11/part-2/badouralix.py
from runners.python import Submission
from collections import Counter

class BadouralixSubmission(Submission):

    def run(self, s):
        # :param s: input in string format
        # :return: solution flag
        steps = Counter()
        group_law = [
            ('n',  's',  ''),
            ('ne', 'sw', ''),
            ('nw', 'se', ''),
            ('n',  'sw', 'nw'),
            ('n',  'se', 'ne'),
            ('s',  'nw', 'sw'),
            ('s',  'ne', 'se'),
            ('ne', 'nw', 'n'),
            ('se', 'sw', 's'),
        ]
        furthest = 0

        for step in s.split(','):
            steps[step] += 1

            reducible = True
            while reducible:
                reducible = False
                for action in group_law:
                    if steps[action[0]] > 0 and steps[action[1]] > 0:
                        reducible = True
                        iterations = min(steps[action[0]], steps[action[1]])
                        steps -= Counter({action[0]:iterations, action[1]:iterations, action[2]:-iterations})
            del steps['']

            furthest = max(furthest, sum(steps.values()))

        return furthest
